---
title: "scDesign3-scATACseq_vignette"
author: 
  - name: Dongyuan Song
    affiliation:
    - Bioinformatics IDP, University of California, Los Angeles
    email: dongyuansong@ucla.edu
output: 
  BiocStyle::html_document:
    self_contained: yes
    toc: true
    toc_float: true
    toc_depth: 2
    code_folding: show
date: "`r doc_date()`"
package: "`r pkg_ver('scDesign3')`"
vignette: >
  %\VignetteIndexEntry{scDesign3-scATACseq_vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    crop = NULL ## Related to https://stat.ethz.ch/pipermail/bioc-devel/2020-April/016656.html
)
```

```{r}
suppressPackageStartupMessages(library(scDesign3))
suppressPackageStartupMessages(library(SingleCellExperiment))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(Signac))
theme_set(theme_bw())
```

## Introduction
In this tutorial, we show how to use scDesign3 to simulate the peak by cell matrix of scATAC-seq data.

## Read in the reference data

The raw data is from the [Signac](https://satijalab.org/signac/articles/pbmc_vignette.html), which is of human peripheral blood mononuclear cells (PBMCs) provided by 10x Genomics. We pre-select the differentially accessible peaks between clusters.

```{r}
example_sce <- readRDS((url("https://www.dropbox.com/s/e75uf5k5voni106/ATAC_sce.rds?raw=1")))
print(example_sce)
```

To save time, we subset 1000 cells.
```{r}
set.seed(123)
example_sce <- example_sce[, sample(seq_len(dim(example_sce)[2]), 1000, replace = FALSE)]
```

## Simulation
Here we choose the Zero-inflated Poisson (ZIP) as the distribution due to its good empirical performance. Users may explore other distributions (Poisson, NB, ZINB) since there is no conclusion on the best distribution of ATAC-seq.

```{r}
set.seed(123)
example_simu <- scdesign3(
    sce = example_sce,
    assay_use = "counts",
    celltype = "cell_type",
    pseudotime = NULL,
    spatial = NULL,
    other_covariates = NULL,
    mu_formula = "cell_type",
    sigma_formula = "1",
    family_use = "zip",
    n_cores = 2,
    usebam = FALSE,
    corr_formula = "cell_type",
    copula = "gaussian",
    DT = TRUE,
    pseudo_obs = FALSE,
    return_model = FALSE,
    nonzerovar = FALSE
  )
```

We also run the TF-IDF transformation.
```{r}
assay(example_sce, "tfidf") <- as.matrix(Signac::RunTFIDF(counts(example_sce)))
simu_sce <- example_sce
counts(simu_sce) <- example_simu$new_count
assay(simu_sce, "tfidf") <- as.matrix(Signac::RunTFIDF(counts(simu_sce)))
```

## Visualization

```{r}
set.seed(123)
compare_figure <- plot_reduceddim(ref_sce = example_sce, 
                                  sce_list = list(simu_sce), 
                                  name_vec = c("Reference", "scDesign3"),
                                  assay_use = "tfidf", 
                                  if_plot = TRUE, 
                                  color_by = "cell_type", 
                                  n_pc = 20)
plot(compare_figure$p_umap)
```
